### PR TITLE
[tests] Cover batching with a small stack

### PR DESCRIPTION
**Pull request by joshlf**
The 85-commit batching test coupled Git's fixed 80-commit limit to
GraphQL's adaptive 64-operation limit. It exercised only success at
those defaults, while creating 85 commits and maintaining 3,595 lines
of output and PR-state snapshots.

Give the strict GitHub fake a configurable operation ceiling and let
test builds lower the Git publication batch size. Replace the large
stack with four commits, forcing GraphQL batches to back off from four
to two and Git publication to make two pushes. Snapshot the complete
operation trace and command output, and keep exact boundary cases in
pure source tests.

---

- 　  #317
- 　  #316
- 👉 #315

**Latest Update:** v13 — [Compare vs v12](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)

<details>
<summary><strong>📚 Full Patch History</strong></summary>

*Links show the diff between the row version and the column version.*

|Version| v12 | v11 | v10 | v9 | v8 | v7 | v6 | v5 | v4 | v3 | v2 | v1 |Base|
|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|:---|
|v13|[v12](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v11](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v10](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v9](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v8](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v13)|
|v12||[v11](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v10](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v9](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v8](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v12)|
|v11|||[v10](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v9](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v8](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v11)|
|v10||||[v9](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v8](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v10)|
|v9|||||[v8](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v9)|
|v8||||||[v7](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v8)|
|v7|||||||[v6](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v7)|
|v6||||||||[v5](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v6)|
|v5|||||||||[v4](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5)|[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v5)|
|v4||||||||||[v3](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4)|[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v4)|
|v3|||||||||||[v2](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3)|[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v3)|
|v2||||||||||||[v1](/joshlf/gherrit/compare/gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2)|[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v2)|
|v1|||||||||||||[Base](/joshlf/gherrit/compare/main..gherrit/Gq6klys556npcrjt7wzqni4m7vmekxk5d/v1)|

</details>
<details>
<summary><strong>⬇️ Download this PR</strong></summary>

######

**Branch**
```bash
git fetch origin refs/heads/Gq6klys556npcrjt7wzqni4m7vmekxk5d && git checkout -b pr-Gq6klys556npcrjt7wzqni4m7vmekxk5d FETCH_HEAD
```

**Checkout**
```bash
git fetch origin refs/heads/Gq6klys556npcrjt7wzqni4m7vmekxk5d && git checkout FETCH_HEAD
```

**Cherry Pick**
```bash
git fetch origin refs/heads/Gq6klys556npcrjt7wzqni4m7vmekxk5d && git cherry-pick FETCH_HEAD
```

**Pull**
```bash
git pull origin refs/heads/Gq6klys556npcrjt7wzqni4m7vmekxk5d
```

</details>

*Stacked PRs enabled by [GHerrit](https://github.com/joshlf/gherrit).*

### CODE REVIEW

**Comment by joshlf** [comment]
@codex review

File: src/pre_push/publication.rs [hunk 1]
```diff
@@ -29,7 +29,22 @@ pub(super) struct PushPlan {
...
+}
+
+#[cfg(gherrit_test)]
+fn push_batch_len() -> usize {
+    std::env::var("GHERRIT_TEST_PUSH_BATCH_LEN").map_or(PUSH_BATCH_LEN, |value| {
+        value
+            .parse::<std::num::NonZeroUsize>()
+            .expect("GHERRIT_TEST_PUSH_BATCH_LEN must be a positive integer")
+            .get()
+    })
```
Comment (joshlf) [inline src/pre_push/publication.rs]:
```suggestion
    items.chunks(if !cfg!(gherrit_test) {
        PUSH_BATCH_LEN
    } else {
        std::env::var("GHERRIT_TEST_PUSH_BATCH_LEN").map_or(PUSH_BATCH_LEN, |value| {
        value
            .parse::<std::num::NonZeroUsize>()
            .expect("GHERRIT_TEST_PUSH_BATCH_LEN must be a positive integer")
            .get()
        })
    }
}
```

**Comment by joshlf** [comment]
@codex review

**Comment by joshlf** [comment]
@codex review
